devicetypes/stelpro/stelpro-ki-zigbee-thermostat.src/stelpro-ki-zigbee-thermostat.groovy

```diff
@@ -0,0 +1,687 @@
+    setOutdoorTemperature(degrees.toDouble(), delay)
+}
+
+def setOutdoorTemperature(Double degrees, Integer delay = 0) {
+    def p = (state.precision == null) ? 1 : state.precision
+    Integer tempToSend
+    def tempToSendInString
+    
+    def celsius = (getTemperatureScale() == "C") ? degrees : (fahrenheitToCelsius(degrees) as Float).round(2)
+
+    if (celsius < 0) {
+        tempToSend = (celsius*100) + 65536
+    }
+    else {
+    	tempToSend = (celsius*100)
+    }
+    tempToSendInString = zigbee.convertToHexString(tempToSend, 4)
+    my_writeAttribute(0x201, 0x4001, 0x29, tempToSendInString, ["mfgCode": "0x1185"])
+}
+
+def increaseHeatSetpoint() {
+    def currentMode = device.currentState("thermostatMode")?.value
+    if (currentMode != "off") {
+		float currentSetpoint = device.currentValue("heatingSetpoint")
+		def locationScale = getTemperatureScale()
+    	float maxSetpoint
+    	float step
+
+    	if (locationScale == "C") {
+        	maxSetpoint = 30;
+        	step = 0.5
+    	}
+    	else {
+        	maxSetpoint = 86
+        	step = 1
+    	}
+
+        if (currentSetpoint < maxSetpoint) {
+            currentSetpoint = currentSetpoint + step
+            quickSetHeat(currentSetpoint)
+        }
+    }
+}
+
+def decreaseHeatSetpoint() {
+    def currentMode = device.currentState("thermostatMode")?.value
+    if (currentMode != "off") {
+        float currentSetpoint = device.currentValue("heatingSetpoint")
+        def locationScale = getTemperatureScale()
+        float minSetpoint
+        float step
+
+        if (locationScale == "C") {
+            minSetpoint = 5;
+            step = 0.5
+        }
+        else {
+            minSetpoint = 41
+            step = 1
+        }
+
+    	if (currentSetpoint > minSetpoint) {
+        	currentSetpoint = currentSetpoint - step
+        	quickSetHeat(currentSetpoint)
+    	}
+    }
+}
+
+def modes() {
+	["heat", "eco", "off"]
+}
+
+def switchMode() {
+    def currentMode = device.currentState("thermostatMode")?.value
+    def lastTriedMode = state.lastTriedMode ?: currentMode ?: "heat"
+    def modeOrder = modes()
+    def next = { modeOrder[modeOrder.indexOf(it) + 1] ?: modeOrder[0] }
+	def nextMode = next(currentMode)
+    def modeNumber;
+    Integer setpointModeNumber;
+    def modeToSendInString;
+
+    if (nextMode == "heat") {
+    	modeNumber = 04
+        setpointModeNumber = 04
+    }
+    else if (nextMode == "eco") {
+    	modeNumber = 04
+        setpointModeNumber = 05
+    }
+    else {
+    	modeNumber = 00
+        setpointModeNumber = 00
+    }
+    if (supportedModes?.contains(currentMode)) {
+        while (!supportedModes.contains(nextMode) && nextMode != "heat") {
+            nextMode = next(nextMode)
+        }
+    }
+    state.lastTriedMode = nextMode
+	modeToSendInString = zigbee.convertToHexString(setpointModeNumber, 2)
+    delayBetween([
+    		"st wattr 0x${device.deviceNetworkId} 0x19 0x201 0x001C 0x30 {$modeNumber}",
```

Comment:
This should use `zigbee.writeAttribute` instead of "st wattr"

Reply:
Ok, just make sure it does work.
Once again, at the time the DTH was written, we couldn't get it to work using zigbee.writeAttribute for this specific message.